src/main/java/com/iopipe/IOpipeService.java

```diff
@@ -242,33 +239,38 @@ public final boolean isEnabled()
 				}
 		}
 		
-		// Keep track of how long execution takes
-		long ticker = System.nanoTime();
-		try
-		{
-			rv = __func.apply(exec);
-		}
+		// Run the function in another thread so that it becomes part of the
+		// given group, this is needed by the profiler plugin
+		__Runner__<R> runner = new __Runner__<R>(exec, __func);
+		Thread runnerthread = new Thread(threadgroup, runner, "main");
 		
-		// An exception or error was thrown, so that will be reported
-		// Error is very fatal, but still report that it occured
-		catch (RuntimeException|Error e)
-		{
-			rt = e;
-			
-			measurement.__setThrown(e);
-		}
+		// Register timeout with this execution number so if execution takes
+		// longer than expected a timeout is generated
+		// Timeouts can be disabled if the timeout window is zero, but they
+		// may also be unsupported if the time remaining in the context is zero
+		__TimeOutWatchDog__ watchdog = null;
+		int windowtime;
+		if ((windowtime = config.getTimeOutWindow()) > 0 &&
+			__context.getRemainingTimeInMillis() > 0)
+			watchdog = new __TimeOutWatchDog__(this, __context,
+				runnerthread, windowtime, coldstarted, exec);
 		
-		// Indicate that execution has finished to the timeout manager
-		// so that it no longer reports timeouts
-		finally
-		{
-			ticker = System.nanoTime() - ticker;
-			if (watchdog != null)
-				watchdog.__finished();
-		}
+		// Start the thread and wait until it dies
+		runnerthread.start();
+		for (;;)
+			try
+			{
+				runnerthread.join();
+				break;
+			}
+			catch (InterruptedException e)
+			{
+				// Ignore
```

Comment:
Should we be ignoring this?

Reply:
In this case yes. Basically methods like `Thread.join()`, `Thread.sleep()`, and `Object.wait()` put the CPU into a sleeping state waiting for something to happen (timeout or a signal in this case). However, Java allows these calls to be interrupted in which they will immediately stop waiting and throw `InterruptedException`, you can then do something like maybe check a flag or something else. The exception is a checked exception so it must always be caught and something done with it (in this case it is ignored and the loop continues). The call `Thread.join()` will block execution until the thread dies for any reason (or if it is already dead it will have no effect), but if it gets interrupted then the thread is probably still alive. We do not need to do any processing if the thread gets interrupted so we can just immediately go back to sleep to allow the other threads more time to run.